### PR TITLE
Esad/fix deadlock block cache

### PR DESCRIPTION
**Pull request by eyusufatik**

### CODE REVIEW

File: crates/ethereum-rpc/src/gas_price/cache.rs [hunk 1]
```diff
@@ -62,8 +64,8 @@ impl<C: sov_modules_api::Context> BlockCache<C> {
         block_number: u64,
         working_set: &mut WorkingSet<C>,
     ) -> EthResult<Option<Rich<Block>>> {
-        let mut number_to_hash = self.number_to_hash.lock().unwrap();
         let mut cache = self.cache.lock().unwrap();
```
Comment (eyusufatik) [inline crates/ethereum-rpc/src/gas_price/cache.rs:67]: see lines 33-34

File: crates/ethereum-rpc/src/gas_price/cache.rs [hunk 2]
```diff
@@ -8,8 +8,10 @@ use sov_modules_api::WorkingSet;
 
 /// Cache for gas oracle
 pub struct BlockCache<C: sov_modules_api::Context> {
-    // Assuming number_to_hash and cache are always in sync
+    /// Assuming number_to_hash and cache are always in sync
+    /// If to be acquired acquire after `cache`
```
Comment (rakanalh) [inline crates/ethereum-rpc/src/gas_price/cache.rs:12]:
```suggestion
    /// Acquire after `cache`
```

File: crates/ethereum-rpc/src/gas_price/cache.rs [hunk 3]
```diff
@@ -8,8 +8,10 @@ use sov_modules_api::WorkingSet;
...
     number_to_hash: Mutex<LruMap<u64, B256, ByLength>>, // Number -> hash mapping
+    /// If to be acquired, acquire before `number_to_hash``
```
Comment (rakanalh) [inline crates/ethereum-rpc/src/gas_price/cache.rs:14]:
```suggestion
    /// Acquire before `number_to_hash``
```

**Review by rakanalh — APPROVED** [review]
LGTM

File: crates/ethereum-rpc/src/gas_price/cache.rs [hunk 4]
```diff
@@ -8,8 +8,10 @@ use sov_modules_api::WorkingSet;
 
 /// Cache for gas oracle
 pub struct BlockCache<C: sov_modules_api::Context> {
-    // Assuming number_to_hash and cache are always in sync
+    /// Assuming number_to_hash and cache are always in sync
+    /// Acquire after `cache`
     number_to_hash: Mutex<LruMap<u64, B256, ByLength>>, // Number -> hash mapping
```
Comment (kpp) [inline crates/ethereum-rpc/src/gas_price/cache.rs:13]: Because it's a (A or B) -> C it's better to keep two lru in 1 struct behind 1 mutex.

**Review by kpp — CHANGES_REQUESTED** [review]
It works but it won't protect us from unintentional bugs in the future

File: crates/ethereum-rpc/src/gas_price/cache.rs (hunk 4 above)
Comment (eyusufatik) [reply crates/ethereum-rpc/src/gas_price/cache.rs:13]: Opening an issue for this

**Review by kpp — APPROVED** [review]

**Pull request merged**